### ISSUE TITLE
mupdf-1.15

### ISSUE THREAD

**Issue by ousia** [pymupdf/PyMuPDF#306]
@JorjMcKie,

I have tried to compile your version 1.14.15 against version 1.15 from _MuPDF_.

I got a bunch of errors and I cannot compile it (I’m using the spec file for _Fedora_).

Is the spec file problematic or is version 1.14.15 incompatible with version 1.15 from _MuPDF_.

Many thanks for your help.

**Comment by JorjMcKie** [comment]
Hello, no this is completely impossible. They have dropped support for embedded files, annotations for other than PDF documents, a separate object type for form fields, etc., etc.
So I have to rewrite major parts of PyMuPDF before this version can be supported.

Von meinem iPhone gesendet

**Comment by JorjMcKie** [comment]
Among other things, I am not willing to give up ``Document.embeddedFile...`` methods.
Because of MuPDF dropping the C code basis for this, I have to create my own C functions to fill that new gap.

I am also still investigating how to best handle that new, more or less equal treatment of annotations and form fields.
It will probably  look like a new ``page`` property ``firstField`` with its own forward chaing to the next form field - independently from the chain ``Page.firstAnnot``.
Both types are now only supported for PDF. Links remain availabl for all doc types.

**Comment by ousia** [comment]
Hi @JorjMcKie,

the removal of embedding capabilities was a really bad move upstream. At least, loosing functionality doesn’t seem extremely wise to me.

I realize that it will take some time to adapt to the new released sources.

I’m closing the issue myself and many thanks for your help.

**Comment by JorjMcKie** [comment]
Hi @ousia,
yes, I agree. Had a bit of trouble to keep emotions under control. The reason they are giving is, that this feature was "hardly used, never free of bugs, and only applicable to a smaal subset of PDFs". They were probably referring to the more advanced portfolio / collections feature with this remark. Embedded files (the only thing used in PyMuPDF) is much more basic and available since PDF 1.3 I believe.

I also feared this might be an early sign for a general slowdown for the development of the MuPDF library, if not worse ... After all, they were also late bringing out this semi-annual version by almost a month.

But there are some improvements as well, so I turned down my concerns again, for the time being. Those improvements are mostly in the area of annotations: again more types supported, also better form field support.
A lot of the same functionality had already been implemented in PyMuPDF (on top of an older MuPDF version), so it won't be noticable in any new v1.15.0.

Bottom line of all of that:
It will take some time before PyMuPDF version 1.15.0. And once it is there, it will not contain exciting new features, presumably.
